### PR TITLE
fix: independent per-leaf noise in shrink-and-perturb

### LINKED ISSUE #161: DQN_Shrink_and_Perturb uses a single shared PRNG key across all parameter leaves
## Summary

`DQN_Shrink_and_Perturb._shrink_and_perturb` (in `src/algorithms/nn/DQN_Shrink_and_Perturb.py` on `main`) reuses one `subkey` for every parameter leaf in `jax.tree_util.tree_map`. Because `jax.random.normal(key, shape)` is deterministic given `(key, shape)`, this means:

- Any two parameter tensors with the **same shape** receive the **identical** perturbation matrix.
- Tensors of different shapes receive noise drawn from the same correlated bit-stream.

This is not the shrink-and-perturb procedure as described in the literature, which requires *independent* Gaussian noise per parameter.

## Buggy code (on `main`, `src/algorithms/nn/DQN_Shrink_and_Perturb.py`)

```python
def _shrink_and_perturb(self, state: AgentState):
    key, subkey = jax.random.split(state.key)

    def sp(p):
        noise = jax.random.normal(subkey, shape=p.shape, dtype=p.dtype)
        return p * state.hypers.shrink_factor + noise * state.hypers.noise_scale

    params = jax.tree_util.tree_map(sp, state.params)
    ...
```

`subkey` is closed over and reused for every leaf, so all leaves of the same shape get the same noise tensor.

## When this was introduced

Commit `7a3fc3c8` ("Implement H&T and S&P"), 2025-09-10. The same pattern was duplicated into `src/rtu_ppo.py`'s shrink-and-perturb path in PR #159.

## Recommended fix (idiomatic `jax.random.split`)

```python
def _shrink_and_perturb(self, state: AgentState):
    key, subkey = jax.random.split(state.key)

    leaves, treedef = jax.tree_util.tree_flatten(state.params)
    leaf_keys = jax.random.split(subkey, len(leaves))

    def sp(k, p):
        noise = jax.random.normal(k, shape=p.shape, dtype=p.dtype)
        return p * state.hypers.shrink_factor + noise * state.hypers.noise_scale

    new_leaves = [sp(k, p) for k, p in zip(leaf_keys, leaves)]
    params = jax.tree_util.tree_unflatten(treedef, new_leaves)
    ...
```

`jax.random.split` is the canonical JAX idiom for deriving N independent keys and is what the JAX docs and `optax`/`flax`/`haiku` recipes use. Each leaf gets a distinct, independently-derived key, so all per-element noise is i.i.d. — exactly what S&P requires.

### Alternative: `tree_map_with_path` + `fold_in`

PR #159 currently uses a `fold_in(subkey, crc32(keypath))` approach. It's correct and has one nice property — adding/reordering leaves doesn't change the noise applied to existing leaves — but it's non-idiomatic, requires an extra hash function, and isn't a pattern in the JAX docs. Prefer `split` unless that order-independence is explicitly desirable; if `fold_in` is kept, add a comment documenting why.

## Scope of impact

`DQN_Shrink_and_Perturb` is referenced in 31 experiment collections under `experiments/`, including:

- `E99-weather`, `E101-two-biome`, `E120-two-biome`, `E123-two-biome`, `E124-diwali`, `E136-big`
- The full `X8…X32` square-wave / weather-wave / sine-biome series

Any S&P numbers from these runs were generated with rank-deficient (shared) noise across same-shape parameter tensors, not the i.i.d. Gaussian noise S&P is supposed to apply.

### Effect on results

- Per-element noise is still zero-mean Gaussian, so the *expected* shrink-and-perturb update is unchanged.
- The covariance structure across leaves is broken: effectively one shared noise tensor (per shape bucket) instead of independent draws.
- For small `noise_scale` the difference is likely modest; for larger scales it could meaningfully bias what S&P looks like as a regularizer.
- If S&P underperformed expectations in any prior comparison, this is a plausible contributing cause.

## Action items

- [ ] Land the fix on `main` as a standalone PR (not bundled inside #159), using `jax.random.split` for consistency with the rest of the JAX ecosystem.
- [ ] Re-run any DQN+S&P results that appear in the paper (especially `X32-ForagaxSquareWaveTwoBiome-v11`) with the fixed implementation before final submission.
- [ ] Audit `src/rtu_ppo.py` shrink-and-perturb path for the same pattern (it's there; PR #159 fixes it but the fix should not be lost if #159 is restructured).
- [ ] Consider adding a unit test that asserts noise drawn for two same-shape parameters under one S&P call is *not* identical, to prevent regression.
- [ ] Decide whether to document in the paper's reproducibility appendix (depends on whether re-runs are feasible before submission).

### PR DESCRIPTION
**Pull request by steventango**
## Summary

`DQN_Shrink_and_Perturb._shrink_and_perturb` closed over a single `subkey` and fed it into `jax.random.normal(subkey, shape=p.shape)` inside a `tree_map`. Since `jax.random.normal(key, shape)` is deterministic given `(key, shape)`, every same-shape parameter leaf received the **identical** noise tensor instead of independent Gaussian draws.

This fix splits `subkey` into one key per leaf with `jax.random.split(subkey, len(leaves))` and threads them through `tree_map`, so each leaf's noise is i.i.d. — matching the canonical `jax.random` idiom.

Fixes #161.

## Notes

- The issue body also calls out the same bug pattern in `src/rtu_ppo.py`, but that path was added in PR #159 and is **not on `main`**. This PR keeps the fix narrowly scoped to what currently exists on `main`. The fix should be carried into `rtu_ppo.py` whenever PR #159 (or its successor) lands.
- Per the issue's "Effect on results" section, prior S&P numbers will need to be re-run; that follow-up is tracked in #161 and is out of scope here.

## Test plan

- [x] `pytest tests/test_dqn_shrink_and_perturb.py -v` — new regression test asserts two same-shape `(4, 4)` leaves receive non-identical perturbations under one S&P call (3 tests pass).
- [x] `pytest tests/ -v` — confirmed no new failures introduced. (17 pre-existing `test_optimizers.py` failures are unrelated — they reference a `param_initializers` kwarg that no longer exists in `selective_weight_reinitialization`.)

### CODE REVIEW

**Comment by steventango** [comment]
## Audit: did this bug actually affect prior results?

Audited all 245 `DQN_Shrink_and_Perturb.json` configs across the 31 affected experiment collections by building each distinct architecture's Haiku param pytree and counting same-shape leaves.

**Only two architectures are used with S&P:**

| Architecture | # configs | Same-shape leaves | Severity |
|---|---|---|---|
| `Forager2Net`, hidden=64 (Conv → Linear(64) → Linear(64)) | 243 / 245 | 2 leaves of `(64,)` (both MLP biases) | cosmetic |
| `ForagerNet` + LN, hidden=64, layers=2 (one X32 sweep cell only) | 2 / 245 | 6 leaves of `(64,)` (2 biases + 2 LN scales + 2 LN offsets) | mild |

**No multi-layer same-width weight matrix duplication exists in any experiment.** Every `(in, out)` shape in `Forager2Net` is unique because input dim differs at each layer (conv kernel `(3,3,3,16)`, flatten `(F,64)`, hidden `(64,64)`, head `(64,4)`). The worst-case scenario from #161 — two `(H, H)` matrices sharing one noise tensor — never occurs. No GRU/RTU architecture is paired with S&P anywhere.

**Effect on results**: weight matrices were correctly perturbed everywhere. Only biases (and a handful of LN params in one sweep cell) occasionally received the same `(64,)` noise vector across 2–6 leaves. Marginal per-element distribution is still i.i.d. Gaussian — what was broken is independence *between* certain bias vectors. For 1-D additive bias perturbations, this barely matters in expectation.

**Re-run scope**: not required for headline numbers in any of the E*/X* runs. The X32 LN sweep cell is mildly suspect but it's a hyperparameter-search artifact, not a production run. Documenting the bug in the paper's reproducibility appendix is probably sufficient instead of a global re-run.

This fix is still correct and worth landing — just for future S&P configs (e.g. anything pairing it with a recurrent core, multi-`(H,H)` MLP, or wider/deeper architecture) where the bug would have substantially mattered.

**Comment by steventango** [comment]
Addressed Copilot review in 2ece65c8:

1. **JIT/tracer concern** — verified with a `jax.jit(...)` smoke test that the production pattern (`tree_unflatten` over `jax.random.split`'s output) traces cleanly. The flagged issue was a false positive (the leading axis of `split`'s output is statically sized via `len(leaves)`, so iteration is concrete at trace time). Added `test_kernel_jits_and_preserves_independence` to cover this code path under JIT going forward.
2. **Tests didn't exercise the jitted path** — same new test wraps the kernel in `jax.jit` and asserts independent noise across same-shape leaves under JIT.
3. **Brittle module-wide source check** — replaced with `inspect.getsource(_shrink_and_perturb)` scoped to the function body, so docstrings/comments elsewhere in the file can't satisfy or break it. The check also accepts both valid implementations (`split` + `tree_map` or `fold_in` + `tree_map_with_path`) so it doesn't lock in our specific fix.

Also added `pythonpath = ["src"]` to `[tool.pytest.ini_options]` so the inspect-based test resolves to the repo's source rather than any unrelated editable install on the active venv.

All 4 tests pass.
